### ISSUE TITLE
automatic parameterized registration of plugin derived permission sets

### ISSUE THREAD

**Issue by PaxJaromeMalues** [SpongePowered/Sponge#3681]
### Is this request specific to one implementation of Sponge?

Yes, only SpongeForge

### Sponge Version

1.12.2 7.4.7

### What are you requesting?

Currently, the only way to get a (semi) complete list of permissions is to create a dummy user and assign all available permissions to it manually or running `lp tree` (in case of luckperms) after the server has been up for a long time.

It would be nice if (maybe also for API7.4) a functionality could be introduced, that automatically fetches all non argument based permission nodes and make them available as a list-file, callrequest or dbentry to the server admins.

**Comment by ejm** [comment]
This likely makes more sense as an issue for SpongeAPI rather than the Sponge repository, as it would be exposing functionality to developers

**Comment by PaxJaromeMalues** [comment]
> This likely makes more sense as an issue for SpongeAPI rather than the Sponge repository, as it would be exposing functionality to developers

damn, you right

**Comment by dualspiral** [comment]
I commented on the API issue (https://github.com/SpongePowered/SpongeAPI/issues/2427), but for completeness...

For what it's worth, this is actually very much an impl request, not an API. The ask here is to enable dumping all known permissions, which is an implementation thing - though there is an element of truth to developers being able to register known permissions (which an API for already exists - `PermissionDescription` on the `PermissionService`).

Thing is, the implementation request is wasted here. Sponge (generally) does not control permissions, that's down to a plugin. Equally, Sponge or other plugins cannot magically know what permissions exist on the server ahead of time without plugins registering the descriptions. The API already exists for plugins to register ahead of time - so for what  you want:

* Permission plugin developers need to implement the `PermissionDescription` part of the permissions service (LuckPerms does this)
* They need to support printing out all known plugins (LuckPerms also does this)
* Other plugins that use permissions need to supply their permissions as `PermissionDescriptions` - most plugins **don't** do this

Sponge has no role to play here - for your problem it sounds like you need to talk to plugin developers who don't use the `PermissionDescription` API, not us, because ultimately, we don't control the permission side of things unless there is no permissions plugin, then we just do op based permissions.

Now, all that said, there are scenarios where permissions _cannot_ be known ahead of time. Take permissions that incorporate the world, `x.y.<worldname>`, where the permission signifies a target world (that is, can't otherwise be determined by standard contexts). How can a plugin know ahead of time that the permission might be needed? It can't! So to magically dump out _all_ permissions is really a fools errand - some permissions cannot be determined ahead of use time.

TL; DR: plugin developers need to register them, we, and permissions plugins, can't magically know everything if they don't tell us. Sponge can't do anything here.

**Comment by PaxJaromeMalues** [comment]
> I commented on the API issue (https://github.com/SpongePowered/SpongeAPI/issues/2427), but for completeness...
> 
> For what it's worth, this is actually very much an impl request, not an API. The ask here is to enable dumping all known permissions, which is an implementation thing - though there is an element of truth to developers being able to register known permissions (which an API for already exists - `PermissionDescription` on the `PermissionService`).
> 
> Thing is, the implementation request is wasted here. Sponge (generally) does not control permissions, that's down to a plugin. Equally, Sponge or other plugins cannot magically know what permissions exist on the server ahead of time without plugins registering the descriptions. The API already exists for plugins to register ahead of time - so for what  you want:
> 
> * Permission plugin developers need to implement the `PermissionDescription` part of the permissions service (LuckPerms does this)
> * They need to support printing out all known plugins (LuckPerms also does this)
> * Other plugins that use permissions need to supply their permissions as `PermissionDescriptions` - most plugins **don't** do this
> 
> Sponge has no role to play here - for your problem it sounds like you need to talk to plugin developers who don't use the `PermissionDescription` API, not us, because ultimately, we don't control the permission side of things unless there is no permissions plugin, then we just do op based permissions.
> 
> Now, all that said, there are scenarios where permissions _cannot_ be known ahead of time. Take permissions that incorporate the world, `x.y.<worldname>`, where the permission signifies a target world (that is, can't otherwise be determined by standard contexts). How can a plugin know ahead of time that the permission might be needed? It can't! So to magically dump out _all_ permissions is really a fools errand - some permissions cannot be determined ahead of use time.
> 
> TL; DR: plugin developers need to register them, we, and permissions plugins, can't magically know everything if they don't tell us. Sponge can't do anything here.

Thank you for the detailed analysis.

So forcing `permissionDescription` ain't a thing as well, because Sponge can not know.
That's kinda a dead end. Sad face.

Is it possible to make the addition of `permissionDescription` for the Devs more attractive? Like priority plugin checks or something?

**Comment by dualspiral** [comment]
In API-8, descriptions are intended to be used as "role templates", basically plugin defined groups that you can apply to other groups etc. In 7, that's not gonna happen now.

That said, how these descriptions are used is up to the permission plugin in question. If they make it compelling to use that function, plugins may follow suit.
